api/app/src/main/kotlin/packit/service/UserService.kt

```diff
@@ -78,4 +81,43 @@ class BaseUserService(
             userRepository.findByUsername(username) ?: throw PackitException("userNotFound", HttpStatus.UNAUTHORIZED)
         return updateUserLastLoggedIn(user, Instant.now())
     }
+
+    override fun addRolesToUser(username: String, roleNames: List<String>)
+    {
+        val (user, rolesToAdd) = getUserAndRolesForUpdate(username, roleNames)
+        if (rolesToAdd.any { user.roles.contains(it) })
+        {
+            throw PackitException("userRoleExists", HttpStatus.BAD_REQUEST)
+        }
+
+        user.roles.addAll(rolesToAdd)
+        userRepository.save(user)
+    }
+
+    override fun removeRolesFromUser(username: String, roleNames: List<String>)
+    {
+        val (user, rolesToRemove) = getUserAndRolesForUpdate(username, roleNames)
+
+        val matchedRolesToRemove = rolesToRemove.map { roleToRemove ->
```

Comment:
Same remark as for permissions on roles - shoulnd't need the map here - once rolesToRemove has been checked you should just be able to use it directly.

Reply:
as per previous pr comment, we need the actual permission... and unsure of benefits of suggested

Reply:
Yep, my bad, sorry